### PR TITLE
[DeadCode][EarlyReturn] Handle RemoveUnusedVariableAssignRector+RemoveAlwaysElseRector

### LINKED ISSUE #7494: Incorrect behavior of RemoveUnusedVariableAssignRector, RemoveAlwaysElseRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/2c6bc559-8a6d-4272-988a-83cc5b9d46f8

```php
<?php

class Get
{
    private $dummyProperty;

    public function __construct()
    {
        $this->dummyProperty = 1;
    }

    private function getLowestPriceItem($value): int
    {
        $newValue = $value + 1;
        if ($value > 100) {
            return $value;
        } else {
            return $newValue;
        }
    }
}
```

### Responsible rules

* `RemoveUnusedVariableAssignRector`
* `RemoveAlwaysElseRector`

## Expected Behavior

The variable definition of `$newValue` will be removed. The Problem only occurs with at least one property assignment in the constructor.

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
class Fixture
{
    private $dummyProperty;

    public function __construct()
    {
        $this->dummyProperty = 1;
    }

    private function getLowestPriceItem($value): int
    {
        $newValue = $value + 1;
        if ($value > 100) {
            return $value;
        } else {
            return $newValue;
        }
    }
}
```

it removes the `$newValue = $value + 1` assignment, while used after `Else_` removed from `If_`, it seems due to check next Node exists is not check that original node has changed.

```diff
     {
-        $newValue = $value + 1;
```

Applied rules:

```php
Rector\DeadCode\Rector\Assign\RemoveUnusedVariableAssignRector;
Rector\EarlyReturn\Rector\If_\RemoveAlwaysElseRector;
```

see https://getrector.org/demo/2c6bc559-8a6d-4272-988a-83cc5b9d46f8
Fixes https://github.com/rectorphp/rector/issues/7494

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @Wohlie

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: tests/Issues/IssueOrIfExplicitBoolCompare/Fixture/fixture.php.inc [hunk 1]
```diff
@@ -28,7 +28,7 @@ class Fixture
      */
     public function run($a = '"', $b)
     {
-        if (! $a) {
+        if ($a === '' || $a === '0') {
```
Comment (samsonasik) [inline tests/Issues/IssueOrIfExplicitBoolCompare/Fixture/fixture.php.inc:31]: This also fix combination between `ExplicitBoolCompareRector` + `ChangeOrIfReturnToEarlyReturnRector` which previously got the bool compare one got skipped and require double run 🎉🎉🎉 

Now, single run can do both

File: src/Rector/AbstractRector.php [hunk 2]
```diff
@@ -377,11 +379,32 @@ private function shouldSkipCurrentNode(Node $node): bool
...
         $nodeTraverser->addVisitor(new ParentConnectingVisitor());
-        $nodeTraverser->traverse([$node]);
+        $nodeTraverser->traverse($nodes);
+    }
+
+    private function connectNodes(array $nodes): void
+    {
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor(new NodeConnectingVisitor());
+        $nodeTraverser->traverse($nodes);
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php:410]: with `NodeConnectingVisitor`, when `refactor()` return multi nodes, they will automatically connected each other 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by TomasVotruba** [comment]
Thanks :partying_face:

**Pull request merged**